### ISSUE TITLE
Filebeat autodiscover stopping too early when kubernetes pod terminates

### ISSUE THREAD

**Issue by cpaton** [elastic/beats#34789]
I am using filebeat with autodiscover within a Kubernetes cluster to capture logs. When a Kubernetes pod terminates filebeat immediately stops reading log entries which can result in log lines at the end of the logs not being published.

Filebeat configuration is setup as

```yaml
filebeat.autodiscover:
  providers:
    - type: kubernetes
      node: ${NODE_NAME}
      resource: pod
      scope: node
      hints.enabled: true
      hints.default_config:
        type: container
        paths:
          - /var/log/containers/*-${data.container.id}.log
        fields_under_root: true          
      include_annotations:
        - app.kubernetes.io/version
        - app-release
```

This issue appears to the same as reported in [Filebeat autodiscover should not stop docker prospectors until it reads all lines · Issue #6694 · elastic/beats (github.com)](https://github.com/elastic/beats/issues/6694) but for the kubernetes input. That was fixed by the introduction of the close_timeout configuration option which is still in place today.

From testing different versions it works as expected up to Filebeat 7.12.1, but is not working as expected from 7.13.0 onwards.

Logs from Filebeat 7.12.1 when a pod terminates

```
2023-02-13T20:27:52.120Z	DEBUG	[autodiscover.pod]	kubernetes/pod.go:129	Watcher Pod update for pod: log-provider-0, status: Running
2023-02-13T20:27:52.854Z	DEBUG	[autodiscover.pod]	kubernetes/pod.go:145	Watcher Pod update (terminating)
2023-02-13T20:28:52.856Z	DEBUG	[autodiscover]	autodiscover/autodiscover.go:253	Got a stop event
2023-02-13T20:28:52.857Z	INFO	input/input.go:136	input ticker stopped
```

Note the 60 second gap between the pod update event being received from Kubernetes as the pod terminates and the stop event being published within filebeat to stop reading the logs

Logs from Filebeat 7.13.0 when a pod terminates

```
2023-02-13T20:26:10.123Z	DEBUG	[autodiscover.pod]	kubernetes/pod.go:145	Watcher Pod update
2023-02-13T20:26:10.128Z	DEBUG	[autodiscover]	autodiscover/autodiscover.go:253	Got a stop event
2023-02-13T20:26:10.129Z	INFO	input/input.go:136	input ticker stopped
```

Note the stop event being published immediately

From some investigation I think it relates to changes in #24742 - Refactor kubernetes autodiscover to avoid skipping short-living pods . Before that change the stop event on pod termination would never be published immediately it was always delayed by the close timeout. Post that change the stop event will be published immediately if any container within the pod has a non-empty running container status.  This is the bit of logic which I think has changed - below is the snippet before the above change

```go
	if pod.GetObjectMeta().GetDeletionTimestamp() != nil {
		p.logger.Debugf("Watcher Pod update (terminating): %+v", obj)
		// Pod is terminating, don't reload its configuration and ignore the event
		// if some pod is still running, we will receive more events when containers
		// terminate.
		for _, container := range pod.Status.ContainerStatuses {
			if container.State.Running != nil {
				return
			}
		}
		time.AfterFunc(p.config.CleanupTimeout, func() { p.emit(pod, "stop") })
		return
	}
```

Now the stop event is not delayed as the pod hasn't terminated, its only in the process of terminating

```go
delay := (flag == "stop" && podTerminated(pod, containers))
```

Note this was originally posted on the discussion forums but did not get any replies - https://discuss.elastic.co/t/filebeat-autodiscover-stopping-too-early-when-kubernetes-pod-terminates/325491

**Comment by StephanErb** [comment]
This could potentially be a duplicate of https://github.com/elastic/beats/issues/34045

**Comment by StephanErb** [comment]
@cpaton do your PODs container a single container or multiple? Your analysis sounds as if the error would only be present if there are multiple containers?

**Comment by cpaton** [comment]
In this example the pod has a single container.

In the first Kubernetes event seen by filebeat when the pod terminates the single containers status has a non-empty running state.  Its this non-empty running state which I think prevents the delay as the pod is deemed to be terminating but not terminated resulting in the input stopping immediately.  The code before #24742 explicitly checked for containers still running in a terminated pod and didn't emit an event

```go
func (p *pod) emit(pod *kubernetes.Pod, flag string) {
	...
	delay := (flag == "stop" && podTerminated(pod, containers))
	...
}

// podTerminating returns true if a pod is marked for deletion or is in a phase beyond running.
func podTerminating(pod *kubernetes.Pod) bool {
	if pod.GetObjectMeta().GetDeletionTimestamp() != nil {
		return true
	}

	switch pod.Status.Phase {
	case kubernetes.PodRunning, kubernetes.PodPending:
	default:
		return true
	}

	return false
}

// podTerminated returns true if a pod is terminated, this method considers a
// pod as terminated if none of its containers are running (or going to be running).
func podTerminated(pod *kubernetes.Pod, containers []*containerInPod) bool {
	// Pod is not marked for termination, so it is not terminated.
	if !podTerminating(pod) {
		return false
	}

	// If any container is running, the pod is not terminated yet.
	for _, container := range containers {
		if container.status.State.Running != nil {
			return false
		}
	}

	return true
}
```

**Comment by calvinbui** [comment]
👍

**Comment by a8950783** [comment]
👍

**Comment by ivoistaken** [comment]
👍

**Comment by hovvsoon** [comment]
👍

**Comment by belimawr** [comment]
Hey folks, have you tried to set [`cleanup_timeout`](https://www.elastic.co/docs/reference/beats/filebeat/configuration-autodiscover#_kubernetes)? I briefly looked at the code and it is used in a different part of the code than the original one and it should have the same effect.

Our code has changed quite a bit since the last comment pointing to a code snippet.

If it is still an issue even after setting `cleanup_timeout` with the latest releases (`v8.19.3`, `v9.0.6`, `v9.1.3`), could you provide the version, the configuration and an example to reproduce it?

**Comment by belimawr** [comment]
@pierrehilbert this is the issue about making Filestream not stop if it is still harvesting a file: https://github.com/elastic/beats/issues/40447

**Comment by belimawr** [comment]
Folks, I've been investigating this for a while now and I cannot reproduce it. As far as I can tell and test, the code works well.

Has any of you been able to reproduce it in a testing scenario, like with kind or minikube?

To continue investigating this I'll need help from you on reproducing.

**Comment by hovvsoon** [comment]
```yaml

 filebeat:
    autodiscover:
        providers:
            - hints:
                default_config:
                    enabled: false
                    paths:
                        - /var/log/containers/*${data.kubernetes.container.id}.log
                    type: container
                enabled: true
              node: ${NODE_NAME}
              type: kubernetes
```

Once Filebeat autodiscover is set up, deploy the YAML below — it runs a .NET app that explicitly logs ASP.NET Core host lifecycle events (startup, graceful shutdown). 



```yaml

kubectl apply -f -<<EOF
apiVersion: apps/v1
kind: Deployment
meta
  name: aspnetlogs
  labels:
    app: aspnetlogs
spec:
  replicas: 5
  selector:
    matchLabels:
      app: aspnetlogs
  template:
    meta
      labels:
        app: aspnetlogs
      annotations:
        co.elastic.logs/enabled: 'true'
        co.elastic.logs/json.ignore_decoding_error: 'true'
        co.elastic.logs/json.message_key: message
    spec:
      containers:
      - name: aspnetlogs
        image: hovvsoon/getting-started-todo-app:hostlifetimelog
        ports:
        - containerPort: 8080
        env:
        - name: ASPNETCORE_ENVIRONMENT
          value: "Production"
        - name: ASPNETCORE_URLS
          value: "http://+:8080"
        resources:
          requests:
            memory: "128Mi"
            cpu: "100m"
          limits:
            memory: "256Mi"
            cpu: "200m"
---
apiVersion: v1
kind: Service
meta
  name: aspnetlogs-service
spec:
  selector:
    app: aspnetlogs
  ports:
  - protocol: TCP
    port: 80
    targetPort: 8080
  type: ClusterIP
EOF
```

Expected: Each Pod should log graceful shutdown messages (e.g., “Application is shutting down…”, “Host stopping…”).

Actual: Some (or sometimes all) Pods have incomplete logs.

**Comment by belimawr** [comment]
Thanks @hovvsoon! I'll try it today.

One quick question, I tried to run the Docker image `hovvsoon/getting-started-todo-app:hostlifetimelog` directly on my machine to get a glimpse of the logs, and once started it just hangs there after logging
```json
{
  "@i": "e947cb35",
  "@m": "Application started successfully - ready to serve requests",
  "@t": "2025-09-18T12:44:08.1107906Z",
  "SourceContext": "HostApplicationLifetimeEventsHostedService"
}
```

To shut down the application I have to press `CTRL+C`.

So for testing it in Kubernetes are you deleting the deployment after a while? Will the application exit after a few minutes? I'm just wondering how you got the pods to stop.

**Comment by belimawr** [comment]
@hovvsoon could you also tell me which version of Filebeat you are using?

**Comment by hovvsoon** [comment]
You can either restart or delete the deployment. You can set Filebeat’s output to console and compare the console logs with `kubectl logs` — you should notice that Filebeat’s output is missing a few log entries triggered by the container termination hooks. I’m using Filebeat version 8.15.1; if this issue persists, I suspect the latest version would have the same problem.

**Comment by belimawr** [comment]
@hovvsoon I believe I figured it out 🎉.

The [`cleanup_timeout`](https://www.elastic.co/docs/reference/beats/filebeat/configuration-autodiscover#_kubernetes) is working as expected, it is not the cause of the problem.

The real root cause is the fact that Kubernetes removes the log files as soon as the containers are removed, and by default the `log` input (`container` in your case) closes the file as soon as it is removed, even if there is data that has not been ingested.

To fix that you need to add `close_removed: false` to your configuration. If you're using the `filestream` input, the option is `close.on_state_change.removed: false`

Here is the final configuration that worked for me:
```yaml
filebeat.autodiscover:
  providers:
    - type: kubernetes
      node: ${NODE_NAME}
      resource: pod
      scope: node
      hints:
        enabled: true
        default_config:
          type: container
          close_removed: false
          enabled: false
          allow_deprecated_use: true # needed in 9.x
          paths:
            - /var/log/containers/*${data.kubernetes.container.id}.log
```

I'll make sure to update our documentation warning about it and update our default configuration to include it.

**Comment by hovvsoon** [comment]
What are the side effects if the file's hold status is not released?

**Comment by belimawr** [comment]
> What are the side effects if the file's hold status is not released?

I'm afraid I don't quite get your question.

What do you mean by a "file's hold status"? Are you referring to keeping the file descriptor open even after it has been deleted?

If a file Filebeat is reading is deleted, but Filebeat keeps the file descriptor open, the OS is not going to free up the resources associated with the file until that file descriptor is closed.

So, if the Filestream input is running with its default configurations, and `close.on_state_change.removed: false`, once Filebeat finishes ingesting the file, it will wait for 5 minutes ([`close.on_state_change.inactive`](https://www.elastic.co/docs/reference/beats/filebeat/filebeat-input-filestream#filebeat-input-filestream-close-inactive)) before closing the file.

**Comment by hovvsoon** [comment]
From the code implementation perspective, `filestream` uses the `taskgroup` mechanism to limit the number of concurrently running goroutines, providing built-in resource control. In contrast, the traditional `log` + `harvester` architecture does not impose any limits on concurrent goroutines. In a Kubernetes environment, if there are a large number of inactive Pods (whose log files are still being monitored but receiving no new data), this may lead to goroutine accumulation and increased scheduling overhead, potentially causing abnormal spikes in CPU usage.

**Comment by belimawr** [comment]
> From the code implementation perspective, `filestream` uses the `taskgroup` mechanism to limit the number of concurrently running goroutines, providing built-in resource control. In contrast, the traditional `log` + `harvester` architecture does not impose any limits on concurrent goroutines. In a Kubernetes environment, if there are a large number of inactive Pods (whose log files are still being monitored but receiving no new data), this may lead to goroutine accumulation and increased scheduling overhead, potentially causing abnormal spikes in CPU usage.

@hovvsoon I understand your performance concern, but I'm a bit puzzled if this is a question, an statement or a feature request 😅.

From the perspective of the original post, I believe this issue will be closed once https://github.com/elastic/beats/pull/46695 is merged.

Spikes in log ingestion can be a challenge in any environment, so I don't believe it is directly related to this issue.

Both, Filestream and the deprecated Log input have mechanism to control the concurrency of log ingestion, however in both cases this limit is applied within one instance of the input, unfortunately there is no way to globally control the concurrency of all Filestream or Log inputs.

When autodiscover is setup as described in this issue (e.g: https://github.com/elastic/beats/issues/34789#issuecomment-3308101321), one instance of the input is created for each container, in other words, one instance of the input is created for each `data.kubernetes.container.id`.

To be able to control the number of active harvesters, you need a single Filestream input, [our documentation](https://www.elastic.co/docs/reference/beats/filebeat/running-on-kubernetes) links to an example manifest that contains a good example, the relevant part is:

```yaml
    filebeat.inputs:
    - type: filestream
      id: kubernetes-container-logs
      paths:
        - /var/log/containers/*.log
      parsers:
        - container: ~
      prospector:
        scanner:
          fingerprint.enabled: true
          symlinks: true
      file_identity.fingerprint: ~
      processors:
        - add_kubernetes_metadata:
            host: ${NODE_NAME}
            matchers:
            - logs_path:
                logs_path: "/var/log/containers/"
```

Then if you want to limit the number of harvesters, you can set `harvester_limit`.

Another option is to set CPU limits in your Kubernets manifest, so a spike in Filebeat's CPU usage won't negatively affect the whole cluster. Just bear in mind that with `close.on_state_change.removed: false` you need enough disk space to keep the files around while Filebeat is shipping them.

Also bear in mind that limiting the number of harvesters might cause Filebeat to never open some files, thus some data loss might happen if they get rotated before Filebeat has a chance to open them.
